File: vs2003_cd01/Program Files/Microsoft Visual Studio .NET 2003/SDK/v1.1/Samples/Technologies/ComponentServices/transactions/CS/TxObj.cs
/*=====================================================================
File:      TxObj.cs

Summary:   Server Code for .NET COM+ Transactions Sample

---------------------------------------------------------------------
This file is part of the Microsoft .NET Framework SDK Code Samples.

Copyright (C) Microsoft Corporation.  All rights reserved.

This source code is intended only as a supplement to Microsoft
Development Tools and/or on-line documentation.  See these other
materials for detailed information regarding Microsoft code samples.

THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
PARTICULAR PURPOSE.
=====================================================================*/

using System;
using System.Data;
using System.Data.SqlClient;
using System.Reflection;
using System.Windows.Forms;
using System.EnterpriseServices;
using System.Runtime.InteropServices;
using System.Runtime.Serialization;

// the ApplicationName attribute specifies the name of the
// COM+ Application that will hold assembly components
[assembly: ApplicationName("TxDemoSvr")]

// the ApplicationActivation.ActivationOption attribute specifies 
// where assembly components are loaded on activation
// Library : components run in the creator's process
// Server : components run in a system process, dllhost.exe
[assembly: ApplicationActivation(ActivationOption.Library)]

// AssemblyKeyFile specifies the name of the strong key
// that will be used to sign the assembly.
// The .snk file was generated with sn.exe
#if BuildBat  // Used when building from the command line
[assembly: AssemblyKeyFile("TxDemoSvrCS.snk")]
#else         // Used when building from within Visual Studio .NET
[assembly: AssemblyKeyFile("..\\..\\TxDemoSvrCS.snk")]
#endif

namespace TxDemoServerCS
{
	// This TxDemoSvr class is going to be hosted within COM+
	// The transaction attribute enables us to do transactions within this class.
	// Requires means that this class needs a transaction, this will be a new 
	// transaction if there isn't one present already, 
	// otherwise it will take part in the existing transaction.
	[Transaction(TransactionOption.Required)] 
	public class TxDemoSvr : ServicedComponent
	{
		private SqlConnection database;

		// First the SQLConnection needs to be setup.
		private SqlConnection Database		
		{
            get
            {
			    if (database == null) database = new SqlConnection(
				    "server=(local)\\NetSDK;Integrated Security=SSPI;database=TxDemoDB");
			    return database;
		    }
    	}

	    // This method returns the "current value" 
	    public int GetCurrentValue()
		{
		    int currentValue = 0;
            try
            {			
                Database.Open();
                SqlCommand sqlCommand = new SqlCommand("select * from currentValue", 
                    Database);
                SqlDataReader sqlDataReader = sqlCommand.ExecuteReader();
                sqlDataReader.Read();
                currentValue = (int)sqlDataReader["currentValue"];
            }
            catch(Exception e)
            {
                throw new CurrentValueNotReadException(e);
            }
            finally
            {
                Database.Close();
            }
			return currentValue ;
		}

		// This method tries to update the "current value", first it updates the 
		// database, then it checks if the value is allowed, if not it throws an 
		// COMException so the [AutoComplete()] Attribute automatically does a 
		// rollback of the transaction.
		[AutoComplete]
		public void AutoCompletePost(int newValue)
		{
            try
            {
                Database.Open();
                SqlCommand sqlCommand = new SqlCommand(@"
																									UPDATE currentValue 
																									SET currentValue=@currentValue
																								",Database);
                sqlCommand.Parameters.Add("@currentValue",SqlDbType.Int,4);
                sqlCommand.Parameters["@currentValue"].Value = newValue;
                try 
                {
                    sqlCommand.ExecuteNonQuery();
                }
                catch(Exception e)
                {
                    throw new DatabaseExecutionException(e);
                }
    		
                if (newValue < 0 || newValue > 10)
                {
                    MessageBox.Show("About to abort the transaction because the new value (" 
                        + newValue + ") is either <0 or >10");
                    throw new ValueOutOfRangeException(newValue);
                }
                else
                {
                    MessageBox.Show("About to commit the transaction");
                }
            }
            finally 
            {
                Database.Close();
            }
		}

		// This method tries to update the "current value", first it updates the 
		// database, then it checks if the value is allowed, if not it calls 
		// ContextUtil.SetAbort to rollback the transaction.
		public void Post(int newValue)
    {
            try
            {
                Database.Open();
                SqlCommand sqlCommand = new SqlCommand(@"
																									UPDATE currentValue 
																									SET currentValue=@currentValue
																								",Database);
                sqlCommand.Parameters.Add("@currentValue",SqlDbType.Int,4);
                sqlCommand.Parameters["@currentValue"].Value = newValue;
                try 
                {
                    sqlCommand.ExecuteNonQuery();
                }
                catch(Exception e)
                {
                    throw new DatabaseExecutionException(e);
                }

                if (newValue<0 || newValue>10)
                {
                    MessageBox.Show("About to abort the transaction because the new value (" 
                        + newValue + ") is either <0 or >10");
                    ContextUtil.SetAbort();
                }
                else
                {
                    MessageBox.Show("About to commit the transaction");
                    ContextUtil.SetComplete();
                }
            } 
            finally 
            {
                Database.Close();
            }
		}
	}
    

		// These are the specialized exceptions this server can throw. 
		// The constructor with the SerializationInfo and the StreamingContext are to 
		// support serialization of the exception as it goes from COM+ to the client.
	    // More information on how to create custom exceptions can be found in the 
		// Exception Handling Whitepaper.
    	
		// CurrentValueNotReadException occurs when there is an error while reading the 
		// current value from the database.
		[Serializable]
    public class CurrentValueNotReadException : Exception
    {
    	public CurrentValueNotReadException(Exception ex) : 
    		base("Unable to get the current value from the database", ex)
    	{
    	}

    	public CurrentValueNotReadException(SerializationInfo info, 
    		StreamingContext context) : base(info, context) 
    	{
    	}
    }

		// DatabaseExecutionException occurs when there is an error while updating 
		// the database
		[Serializable]
    public class DatabaseExecutionException : Exception
    {
        
    	public DatabaseExecutionException(Exception ex) : 
				base("Error while executing database commands", ex)
    	{
    	}
        
    	public DatabaseExecutionException(SerializationInfo info, 
    		StreamingContext context) : base(info, context) 
    	{
    	}
    }

		// AbortTransactionException occurs when an AutoComplete attribute is used 
		// and the transaction needs to be aborted
		[Serializable]
    public class ValueOutOfRangeException : Exception
    {
    	public ValueOutOfRangeException(int newValue) : 
    		base("the new value (" + newValue + ") is either <0 or >10")
    	{
    	}

    	public ValueOutOfRangeException(SerializationInfo info, 
    		StreamingContext context) : base(info, context) 
    	{
    	}
    }

}

File: vs2003_cd01/Program Files/Microsoft Visual Studio .NET 2003/SDK/v1.1/Samples/Technologies/ComponentServices/transactions/VB/TxObj.vb
'=====================================================================
'  File:      TxObj.vb
'
'  Summary:   Server Code for .NET COM+ Transactions Sample
'
'---------------------------------------------------------------------
'  This file is part of the Microsoft .NET Framework SDK Code Samples.
'
'  Copyright (C) Microsoft Corporation.  All rights reserved.
'
'This source code is intended only as a supplement to Microsoft
'Development Tools and/or on-line documentation.  See these other
'materials for detailed information regarding Microsoft code samples.
'
'THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
'KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'PARTICULAR PURPOSE.
'=====================================================================

Imports System
Imports System.Data
Imports System.Data.SqlClient
Imports System.Reflection
Imports System.Windows.Forms
Imports System.EnterpriseServices
Imports System.Runtime.InteropServices
Imports System.Runtime.Serialization


' the ApplicationName attribute specifies the name of the
' COM+ Application that will hold assembly components
<Assembly: ApplicationName("TxDemoSvr")> 

' the ApplicationActivation.ActivationOption attribute specifies 
' where assembly components are loaded on activation
' Library : components run in the creator's process
' Server : components run in a system process, dllhost.exe
<Assembly: ApplicationActivation(ActivationOption.Library)> 

' AssemblyKeyFile specifies the name of the strong key
' that will be used to sign the assembly.
' The .snk file was generated with sn.exe
<Assembly: AssemblyKeyFile("TxDemoSvrVB.snk")> 

Namespace TxDemoServerVB

    'This TxDemoSvr class is going to be hosted within COM+
    'The transaction attributes enables us to do transactions within the class.
    <Transaction(TransactionOption.Required)> _
    Public Class TxDemoSvr
        Inherits ServicedComponent

        Private sqlConnection As SqlConnection


        'First the SQLConnection needs to be setup.
        Private ReadOnly Property Database as SqlConnection
        	Get
            	if sqlConnection Is Nothing then sqlConnection = New sqlConnection _
            		("server=(local)\NetSDK;Integrated Security=SSPI;" & _
            		"database=TxDemoDB")
            	Database = sqlConnection
            End Get
        End Property

        'This method returns the "current value" 
        Public Function GetCurrentValue() As Integer
        	Dim currentValue As Integer = 0
            Try            
	            Database.Open()
            	Dim sqlCommand As New sqlCommand("select * from currentValue", _
            		Database)
            	Dim sqlDataReader As sqlDataReader = sqlCommand.ExecuteReader()
            	sqlDataReader.Read()
            	currentValue = CInt(sqlDataReader("currentValue"))
            	sqlConnection.Close()
            Catch ex as Exception           	
				throw new CurrentValueNotReadException(ex)
			Finally
                Database.Close()
            End Try
            GetCurrentValue = currentValue
        End Function
    	
        'This method tries to update the "current value", first it updates the
        'database, then it checks if the value is allowed, if not it throws an 
        'COMException so the <AutoComplete()> Attribute automatically does a 
        'rollback of the transaction.
        <AutoComplete()> _
        Public Sub AutoCompletePost(ByVal newValue As Integer)
        	Try
				Database.Open()
				Dim sqlCommand As New sqlCommand("UPDATE currentValue SET " & _
					"currentValue=@currentValue", Database)
				sqlCommand.Parameters.Add("@currentValue", SqlDbType.Int, 4)
				sqlCommand.Parameters("@currentValue").Value = newValue
				Try
                    sqlCommand.ExecuteNonQuery()
    			Catch ex As Exception
    				Throw New DatabaseExecutionException(ex)
	    		End Try
    		
                If newValue < 0 Or newValue > 10 Then
                    MessageBox.Show(("About to abort the transaction because "& _
                	    "the new value (" & newValue & ") is either <0 or >10"))
                    Throw New ValueOutOfRangeException(newValue)
                Else
                    MessageBox.Show("About to commit the transaction")
                End If
            Finally
                Database.Close()
            End Try
        End Sub


        'This method tries to update the "current value", first it updates the 
        'database, then it checks if the value is allowed, if not it calls 
        'ContextUtil.SetAbort to rollback the transaction.
        Public Sub Post(ByVal newValue As Integer)
        	Try        	
            	Database.Open()
            	Dim sqlCommand As New sqlCommand("UPDATE currentValue SET " & _
            		"currentValue=@currentValue", Database)
            	sqlCommand.Parameters.Add("@currentValue", SqlDbType.Int, 4)
	            sqlCommand.Parameters("@currentValue").Value = newValue
	            Try
                    sqlCommand.ExecuteNonQuery()
	            Catch ex as Exception
    	        	Throw New DatabaseExecutionException(ex)
	            End Try
			    If newValue < 0 Or newValue > 10 Then
				    MessageBox.Show(("About to abort the transaction because "& _
					    "the new value (" & newValue & ") is either <0 or >10"))
				    ContextUtil.SetAbort()
                Else
                    MessageBox.Show("About to commit the transaction")
                    ContextUtil.SetComplete()
                End If
            Finally
                Database.Close()
            End Try
        End Sub
    End Class
    
	'These are the specialized exceptions this server can throw. 
	'The constructor with the SerializationInfo and the StreamingContext 
	'are to support serialization of the thread as it goes from COM+ to the 
	'client.

	'CurrentValueNotReadException occurs when there is an error while reading 
	'the current value from the database

	<Serializable()> _
	Public Class CurrentValueNotReadException 
        Inherits Exception
		Sub New(ex as Exception)
			MyBase.New("Unable to get the current value from the database", ex)
		End Sub
    	
		Sub New(info as SerializationInfo, context as StreamingContext)  
			MyBase.New(info, context) 
		End Sub
	End Class

	'DatabaseExecutionException occurs when there is an error while updating 
	'the database	
	<Serializable()> _
	Public Class DatabaseExecutionException 
        Inherits Exception
		Sub New(ex as Exception)
			MyBase.New("Error while executing database commands", ex)
		End Sub
    	
		Sub New(info as SerializationInfo, context as StreamingContext)  
			MyBase.New(info, context) 
		End Sub
	End Class
    
	'AbortTransactionException occurs when an AutoComplete attribute is used and 
	'the transaction needs to be aborted
    <Serializable()> _
    Public Class ValueOutOfRangeException
        Inherits Exception
        Sub New(ByVal newValue As Integer)
            MyBase.New("the new value (" & newValue & ") is either <0 or >10")
        End Sub

        Sub New(ByVal info As SerializationInfo, ByVal context As StreamingContext)
            MyBase.New(info, context)
        End Sub
    End Class

End Namespace